wallet: improve rescan performance 640%

pstratem opened the pull request:
The BIP158 compact block filters make a compromise between performance and security by re-keying the filter for each block.  That means every block requires re-hashing and re-sorting for every script pub key before matching against the filter.

To improve on the performance we can simply generate filters using per index random keys (rather than per filter keys).  We don't share these filters with any attacker so there is no security risk.

I don't use synthetic benchmarks for this as we have only a single data set we're ever processing.

On an arbitrary server:
No filters: 132 minutes
Block filters: 45 minutes
Wallet filters: 7 minutes

Commit:
Implement WalletFilterIndex.

This index is very similar to the BIP158 compact block filter indexes with the
key difference that the siphash key is random per node rather than per block.

The static key makes it possible to build a pre-hashed, pre-sorted query set
for very fast scanning of block filters.

Commit:
Refactor GCSFilter::HashToRange into HashElement and RangeHashedElement.

Commit:
Introduce GCSFilter::QuerySet a cached hash query set for GCSFilter::MatchAny

Caching the hashed and sorted elements allows for a substantially faster
GCSFilter::MatchAny. We will use this later for faster wallet rescans.

Commit:
Introduce GCSFilter::MatchAny that supports QuerySet

Commit:
Add walletfilterindex chain interfaces.

Commit:
Use wallet filter during rescan if available.

0xB10C commented:
While a 640% performance improvement is nice, a PR of this size could probably include a bit more details explaining what you've changed and why you've changed it in the OP and the commit messages. This makes it easier for reviewers to approach and review your PR.

mzumsande reviewed (commented):
Could you elaborate about when it would be advisable use this new feature? As far as I understand, the cost for this optimization is that we'd need to build and store a new index for rescanning - which will also take time and consume permanent disk space (how much?). Am I right to think that activating this would mostly make sense in specific situations where you know that you will need to rescan frequently in the future?

aureleoules commented:
> which will also take time and consume permanent disk space (how much?).

On mainnet is takes 8.4G currently.

theStack reviewed (commented):
Agree with previous reviewers that a more detailled PR description would be very helpful. For providing more systematic benchmarking results, you could use the approach from #25957 (see table in the PR description and the linked [functional test benchmark](https://github.com/theStack/bitcoin/blob/fast_rescan_functional_test_benchmark/test/functional/pr25957_benchmark.py)) as a starting point. IIRC you mentioned on IRC that the speedup only show on a quite high element set size, i.e. a large number of scriptPubKeys in the wallet's keypool -- would be very interesting to also benchmark this to get a feeling what type of wallet users (w.r.t. to tx frequency) would benefit the most from this optimization.

> Am I right to think that activating this would mostly make sense in specific situations where you know that you will need to rescan frequently in the future?

Probably another potential (non-wallet) use-case for this index would be to speed up the `scanblocks` RPC.

Commit:
rpc: doc: mention rescan speedup using walletfilterindex=1 in affected wallet RPCs

pstratem commented:
@0xB10C I've reorganized the commits to be easier to understand and written a small explanation.
@mzumsande Building the wallet filter index is optional, user with many keys and/or many wallets would greatly benefit from the performance improvement.
@theStack I did a benchmark against no filters and bip158 filters on the mainnet chain specifically because it's really the only benchmark that matters.

pstratem commented:
@theStack a more useful performance benchmark is that im looking into improving the filter decoder since that's about 50% of the cpu time

theStack commented:
@pstratem: The main idea of the benchmark would be to determine the performance dependent of the wallet's keypool size. It's rather unlikely that a freshly created wallet (creating 8000 keys with default settings) shows the same speed-up as one that is in heavily in use daily and has hundreds of thousands keys in the pool, e.g. an exchange. At the very least, it would be interesting to know the keypool size of the wallet you used for the results shown in the PR description.

pstratem commented:
@theStack the wallet I used had 10k keys, with more keys it will be slower since it does till have to do some processing for each key for each block.

brunoerg commented:
Is it possible to see this performance improvement in any functional test?

Sjors commented:
I agree with others above that this pull requests needs more explanation. The name "wallet" is also a bit confusing, since the `basic` block filter also used for wallets.

Can this be implemented with a new type of block filter, so that adding the index is a matter of `-blockfilterindex=fast`? That would probably make this PR smaller.

pstratem commented:
@Sjors no it cannot
> Can this be implemented with a new type of block filter, e.g. call it FAST, so that adding the index is a matter of -blockfilterindex=fast? That would probably make this PR smaller.

pstratem commented:
@Sjors I'm not sure what else needs explaining, but I'm happy to explain things if i know what's confusing.

pstratem commented:
Let me clarify some, basically I don't think the BlockFilter/BlockFilterIndex interface can be reasonably changed to be fast. There's a few issues (at least one of which is my fault).

The BlockFilter and BlockFilterIndex interfaces assume that the GCSFilter::Params can be generated just from the filter type and the block_hash.  I previously tried changing these interfaces so that they didn't make this assumption, but I found that to involve a lot of changes that were frankly just confusing.

The BlockFilter serialization matches the BIP157/158 serialization, but isn't the serialization that I want for the wallet rescan index.  Specifically the index ends up writing the block_hash twice and the filter_type to the flat file, both of which are implied when looking up the filter in the first place. (They waste disk space and decode time).

Lastly the BlockFilterIndex appends a sha256 hash to the end to check for disk corruption, this would be a significant cost (indeed the siphash that I'm using now is 10% of the total runtime, I'm going to replace it with a crc32c).

achow101 reviewed (commented):
`getindexinfo` needs to be updated to include the wallet filter index.

I'm not entirely convinced that the extra effort to implement and maintain an index of filters for wallet use only is actually worth the benefit. While this does significantly improve the rescan time, it's not clear how often people rescan to make this worthwhile. Rescanning shouldn't be done all that frequently, so this might not end up being all that useful? It'd certainly be less of an issue if the index could be deduplicated with the block filter index.

achow101 commented on src/Makefile.am:
```diff
@@ -463,6 +463,7 @@ endif
 libbitcoin_wallet_a_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) $(BOOST_CPPFLAGS) $(BDB_CPPFLAGS) $(SQLITE_CFLAGS)
 libbitcoin_wallet_a_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
 libbitcoin_wallet_a_SOURCES = \
+  blockfilter.cpp \
```
In 050716c48d703424dc94fde9d7f987b55cbff6a9 "Use wallet filter during rescan if available."

I feel like there's a better way to do this, but I'm not familiar enough with build systems to suggest one. Perhaps @fanquake might have an idea?

pstratem replied:
yes this felt kind of wrong but I didn't see another way to achieve the result. suggestions definitely wanted.

pstratem commented:
@achow101 These filters can be used for everything we use the other block filters for, I've only implemented the wallet query because that's the biggest win I see, but there's no reason they couldn't be used everywhere else (where there might be similar speed ups available).

pstratem commented on src/blockfilter.cpp:
```diff
@@ -24,6 +24,32 @@ static const std::map<BlockFilterType, std::string> g_filter_types = {
+    for (const uint64_t& hashed_element : m_hashed_elements) {
+        ranged_elements.push_back(RangeHashedElement(hashed_element, F));
+    }
+    return ranged_elements;
```
m_hashed_elements needs to be sorted before it's used in the comparison, whether it's before or after being ranged doesn't change the result, but does change the runtime (the range function doesn't change the sort order of the list).

by doing the sort before ranging we only have to do it once per unique set of elements, which is a performance win.

it looks like i did mess up the initial commit, it's fixed in the later commit that makes sorting explicit, I can squash most of that into the initial commit.

pstratem replied:
I've broken out this change into it's own commit with an explanation.

pstratem commented on src/index/walletfilterindex.cpp:
```diff
@@ -0,0 +1,265 @@
+/** The pre-allocation chunk size for fltr?????.dat files */
+constexpr unsigned int FLTR_FILE_CHUNK_SIZE = 0x100000; // 1 MiB
+
+// value for M explained https://github.com/sipa/writeups/tree/main/minimizing-golomb-filters
+constexpr uint8_t WALLET_FILTER_P = 19;
+constexpr uint32_t WALLET_FILTER_M = 784931; // 1.497137 * pow(2,WALLET_FILTER_P);
```
It might make sense to change these values from the BASIC filter values, changing P =32 improves the false positive rate from 1 in 784,931 to 1 in 6,430,154,452 for a trade off of 68% larger filters.

It also might make sense to make it a configuration option when the filters are generated, the value could easily be stored in the index leveldb instead of being a constant.

pstratem commented on src/index/walletfilterindex.h:
```diff
@@ -0,0 +1,54 @@
+
+static constexpr bool DEFAULT_WALLETFILTERINDEX{false};
+
+class WalletFilterIndex final : public BaseIndex
```
WalletFilterIndex::CustomCommit, WalletFilterElements, and DBHashKey are all candidates.

WalletFilterElements and DBHashKey I can certainly refactor to consolidate the logic.

The CustomCommit method though seems like it would end up adding a bunch of complexity to reuse with very little benefit.

pstratem commented on test/functional/wallet_faster_rescan.py:
```diff
@@ -0,0 +1,104 @@
+NUM_BLOCKS = 6       # number of blocks to mine
+
+
+class WalletFasterRescanTest(BitcoinTestFramework):
```
I think I can do that, but I haven't written many of the functional tests so that might take some trial and error.

pstratem commented on src/blockfilter.cpp:
```diff
@@ -158,9 +153,9 @@ bool GCSFilter::MatchInternal(const uint64_t* element_hashes, size_t size) const
         while (true) {
             if (hashes_index == size) {
                 return false;
-            } else if (element_hashes[hashes_index] == value) {
+            } else if (RangeHashedElement(element_hashes[hashes_index]) == value) {
                 return true;
-            } else if (element_hashes[hashes_index] > value) {
+            } else if (RangeHashedElement(element_hashes[hashes_index]) > value) {
```
No the truncation doesn't change the sort order, the fast range function is just (x*n) >> 64, the order of which is always going to be the same as the order of x.

Commit:
Simplify filter lookup interface to only require the block_hash.

Commit:
Eliminate BuildHashedSet to make logic more flexible for future changes.

Commit:
Move the RangeHashedElement call into MatchInternal.

Previously we hashed, ranged, and then sorted. Now we're hashing, sorting,
and then ranging. This is ok because the order of the hashed elements
will always match the order of the ranged hashed elements.

This avoids a copy of the hashed elements being made.

josibake commented:
Concept -0

While this may be very useful for certain wallet usage patterns, this doesn't seem generally useful for _all_ wallets, which makes it hard to justify the review/complexity/maintenance burden.


> @achow101 These filters can be used for everything we use the other block filters for, I've only implemented the wallet query because that's the biggest win I see, but there's no reason they couldn't be used everywhere else (where there might be similar speed ups available).

Wallet rescans and serving BIP158 filters are the only use cases for the block filter index that I'm aware of. If you have other examples, it would be great to enumerate them.

> maybe I should change the name of the index, FixedBlockFilterIndex ?

Rather than make a new index, if there are ways to improve the existing `BlockFilter` index, I think that would be much better as we would be improving two existing use cases without needing to maintain a new index.

Overall, while there is clearly an improvement over using the existing `BlockFilter` index for wallet rescans, I think wallet rescans with the existing index are "good enough" and adding an extra index provides diminishing returns.

furszy commented:
> Rather than make a new index, if there are ways to improve the existing BlockFilter index, I think that would be much better as we would be improving two existing use cases without needing to maintain a new index.

@josibake, orthogonal topic but check #26966 and #27006 :).
